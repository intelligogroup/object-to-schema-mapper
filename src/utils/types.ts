type SomeObj = { [key: string]: any }

type InstructionStrategies = {
    [key: string]: (object: SomeObj, instruction) => SomeObj;
}

interface PredefinedTransformation {
    transformationArgs?: any 
    transformation: 'toUpperCase'
    | 'toLowerCase'
    | 'titleCase'
    | 'toDate'
    | 'stringToArray'
    | 'arrayToString'
}

type TreeLeaf = [string, Transform['target'][]]

type Transform = {
    source: string,
    target: {
        path: string,
        priority?: number,
        defaultValue?: any,
        conditionalValue?: {
            targetPathCondition: string,
            value: any
        }
        predefinedTransformations?: PredefinedTransformation[]
    }
}

type ValType = 'ARRAY'
    | 'STRING'
    | 'BOOLEAN'
    | 'OBJECT'
    | 'NULL'
    | 'NUMBER'
    | 'UNDEFINED'

type ValArchtype = 'SIMPLE'
    | 'EMPTY'
    | 'ARRAY'
    | 'OBJECT'

export {
    SomeObj,
    Transform,
    InstructionStrategies,
    ValType,
    ValArchtype,
    TreeLeaf,
};
