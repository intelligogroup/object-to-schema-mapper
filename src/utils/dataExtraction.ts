type Path = [string];

const getNestedValue = (obj: any) => (path: Array<string>): any => {
    if (obj == null) {
        return null;
    }
    const [key, ...nextKeys] = path;
    return nextKeys.length > 0
        ? getNestedValue(obj[key])(nextKeys)
        : obj[key]
            ? obj[key]
            : null;
};

const isNonEmptyString = (str: string) => str.trim().length > 0;

const getFromArray = (arr: Array<any>) => (path: Path): Array<any> => {
    if (arr == null) {
        return [];
    } else if (path != null) {
        return arr.map(obj => getNestedValue(obj)(path));
    } else {
        return arr.reduce((acc, val) => acc.concat(val), []);
    }
};

const getSeveral = (obj: any) => (pathArray: Array<Path>): any => {
    if (obj == null) {
        return [];
    } else if (Array.isArray(obj)) {
        return pathArray.map(path => getFromArray(obj)(path));
    } else {
        return obj == null ? [] : pathArray.map(path => getNestedValue(obj)(path));
    }
};

const progressNestedObject = (result: any, operation: any) => {
    if (result == null || result.length == 0) {
        return result;
    }
    const { fn, path } = operation;
    const ans = fn(result)(path);
    if (Array.isArray(ans)) {
        return ans.filter(obj => obj != null);
    }
    return ans;
};

const NestedOfObject = function (root: object) {
    this.operations = [];
    this.isArray = false;

    this.addOperation = ({ fn, path }: { fn: Function, path: Path }) => {
        this.operations = [...this.operations, { fn, path }];
    }

    this.inPath = (path: Path) => {
        this.addOperation({ fn: getNestedValue, path });
        return this;
    }

    this.inArray = (path: Path) => {
        this.isArray = true;
        this.addOperation({ fn: getFromArray, path });
        return this;
    }

    this.several = (pathArray: Array<Path>) => {
        this.addOperation({ fn: getSeveral, path: pathArray });
        return this;
    }

    this.filterNullsInArray = (arr: Array<any>) =>
        arr.map(
            inner => Array.isArray(inner)
                ? inner.filter(el => el != null)
                : inner
        ).filter(el => el != null);

    this.get = () => {
        const ans = this.operations.reduce(progressNestedObject, root);
        if (Array.isArray(ans)) {
            return this.filterNullsInArray(ans);
        }
        return ans;
    }

    this.exist = () => {
        if (this.isArray) {
            const ans = NestedOf(this.get()).inArray().get();
            return ans != null && ans.length > 0;
        }
        return this.get() != null;
    }
}

function NestedOf(root: object): any {
    return new NestedOfObject(root);
}

export { NestedOf, };
