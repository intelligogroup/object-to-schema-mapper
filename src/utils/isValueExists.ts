export default function isValueExists(value: any): boolean {
    return value !== undefined && value !== null;
}
