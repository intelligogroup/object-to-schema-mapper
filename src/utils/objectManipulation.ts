type Path = string[];

const getNestedValue = (obj: any) => (path: string[]): any => {
    if (obj == null) {
        return null;
    }
    const [key, ...nextKeys] = path;
    return nextKeys.length > 0
        ? getNestedValue(obj[key])(nextKeys)
        : obj[key]
            ? obj[key]
            : null;
};

const setNestedValue = (obj: any, value: any) => (path: string[]): any => {
    if (obj == null) {
        return null;
    }
    const [key, ...nextKeys] = path;
    if (!Object.keys(obj).includes(key)) {
        obj.key = null;
    }
    return nextKeys.length > 0
        ? setNestedValue(obj[key], value)(nextKeys)
        : Object.assign(obj, { [key]: value });
}

const isNonEmptyString = (str: string) => str.trim().length > 0;

const getFromArray = (arr: any[]) => (path: Path): any[] => {
    if (arr == null) {
        return [];
    } else if (path != null) {
        return arr.map(obj => getNestedValue(obj)(path));
    } else {
        return arr.reduce((acc, val) => acc.concat(val), []);
    }
};

const getSeveral = (obj: any) => (pathArray: Path[]): any => {
    if (obj == null) {
        return [];
    } else if (Array.isArray(obj)) {
        const keysToExtract = pathArray.reduce((acc, elm) => [...acc, ...elm], []);
        return obj
            .map(source =>
                Object
                    .keys(source)
                    .filter(key => keysToExtract.includes(key))
                    .reduce((acc, key) => ({ ...acc, ...{ [key]: source[key] } }), {})
            )
    } else {
        return obj == null ? [] : pathArray.map(path => getNestedValue(obj)(path));
    }
};

const progressNestedObject = (result: any, operation: any) => {
    if (result == null || result.length == 0) {
        return result;
    }
    const { fn, path } = operation;
    const ans = fn(result)(path);
    if (Array.isArray(ans)) {
        return ans.filter(obj => obj != null);
    }
    return ans;
};

const NestedOfObject = function (root: object) {
    this.operations = [];
    this.isArray = false;

    this.addOperation = ({ fn, path }: { fn: Function, path: Path }) => {
        this.operations = [...this.operations, { fn, path }];
    }

    this.inPath = (path: Path) => {
        this.addOperation({ fn: getNestedValue, path });
        return this;
    }

    this.inArray = (path: Path) => {
        this.isArray = true;
        this.addOperation({ fn: getFromArray, path });
        return this;
    }

    this.several = (pathArray: Path[]) => {
        this.addOperation({ fn: getSeveral, path: pathArray });
        return this;
    }

    this.filterNullsInArray = (arr: any[]) =>
        arr.map(
            inner => Array.isArray(inner)
                ? inner.filter(el => el != null)
                : inner
        ).filter(el => el != null);

    this.get = () => {
        const ans = this.operations.reduce(progressNestedObject, root);
        if (Array.isArray(ans)) {
            return this.filterNullsInArray(ans);
        }
        return ans;
    }

    this.exist = () => {
        if (this.isArray) {
            const ans = NestedOf(this.get()).inArray().get();
            return ans != null && ans.length > 0;
        }
        return this.get() != null;
    }
}

function NestedOf(root: object): any {
    return new NestedOfObject(root);
}

export { NestedOf, };
