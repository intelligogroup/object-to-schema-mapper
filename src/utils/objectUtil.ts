import { ValType, ValArchtype } from './types'

function resolveValueArchtype(value: any): ValArchtype {
    const type = resolveType(value);
    switch (type) {
        case 'NULL':
        case 'UNDEFINED':
            return 'EMPTY';
        case 'NUMBER':
        case 'BOOLEAN':
        case 'STRING':
            return 'SIMPLE';
        default:
            return type;
    }
}

function resolveType(value: any): ValType {
    if (value == null) {
        return 'NULL';
    }
    else if (Array.isArray(value)) {
        return 'ARRAY';
    }
    else if (typeof value == 'string') {
        return 'STRING';
    }
    else if (typeof value == 'boolean') {
        return 'BOOLEAN';
    }
    else if (typeof value == 'object') {
        return 'OBJECT';
    }
    else if (
        Number.isInteger(value)
        || (!Number.isNaN(value) && Number.isInteger(Math.floor(value)))
    ) {
        return 'NUMBER';
    }
    return 'UNDEFINED';
}

export { resolveType, resolveValueArchtype };
