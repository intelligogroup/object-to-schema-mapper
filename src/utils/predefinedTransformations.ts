import { applyToOneOrMany } from './transform';
import objectPath, { get } from 'object-path';
import * as chrono from 'chrono-node';


const toLowerCase = applyToOneOrMany<string, string>(str => str.toLowerCase());
const toUpperCase = applyToOneOrMany<string, string>(str => str.toUpperCase());
const titleCase = applyToOneOrMany<string, string>(titleCaseTransformer)
const companyNameFormat = applyToOneOrMany<string, string>(companyNameTransformer)

function titleCaseTransformer(str) {
    const keepCapitalizedWords = [
        "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X"
    ];

    const makeCapitalizedWords = [
        'AK', 'AL', 'AR', 'AS', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'GU', 'HI', 'IA',
        'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MP', 'MS', 'MT',
        'NC', 'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'PR', 'RI', 'SC', 'SD',
        'TN', 'TX', 'UT', 'VA', 'VI', 'VT', 'WA', 'WI', 'WV', 'WY', "UK", "US", "USA"
    ];

    if (!str) {
        return str;
    }

    return str.trim()
        .split(/\s+/)
        .map(word => {
            const cleanedWord = word.replace(/^[^a-zA-Z]+|[^a-zA-Z]+$/g, '');

            if (keepCapitalizedWords.includes(cleanedWord)) {
                return word;
            }

            const cleanedWordIWithEscapeChars = cleanedWord.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
            const capitalizeWordPattern = new RegExp(`^${cleanedWordIWithEscapeChars}$`, 'i');
            if (makeCapitalizedWords.some(w => capitalizeWordPattern.test(w))) {
                return word.toUpperCase();
            }

            let firstLetterIndex = word.search(/[a-zA-Z]/);
            if (firstLetterIndex !== -1) {
                return word.slice(0, firstLetterIndex) +
                    word.charAt(firstLetterIndex).toUpperCase() +
                    word.slice(firstLetterIndex + 1).toLowerCase();
            }

            return word;
        })
        .join(' ');
}


const toDate = applyToOneOrMany<string, string>(str => new Date(str).toISOString());
const stringToArray = applyToOneOrMany<string, string[]>((str, separator) => str.split(separator as string));

function fromTracerDate(tracerDate: number) {

    if (!tracerDate) {
        return;
    }

    const year = String(tracerDate).substr(0, 4);
    const month = String(tracerDate).substr(4, 2);
    const day = String(tracerDate).substr(6, 2);

    return new Date(`${year}-${month}-${day}`).toString();
}


function tracerPropertyDocumentType(documentType: string) {

    let value;

    switch (documentType) {

        case 'G': {
            value = 'Grant Deed/Deed of Trust';
            break;
        }

        case 'H': {
            value = 'High Liability';
            break;
        }

        case 'Q': {
            value = 'Quitclaim/Deed of Trust';
            break;
        }

        case 'W': {
            value = 'Warranty Deed/Deed of Trust';
            break;
        }

        default: {
            value = documentType
            break;
        }
    }

    return value;
}

function tracerPropertyTransferType(documentType: string) {

    let value;

    switch (documentType) {
        case 'R': {
            value = 'Resale';
            break;
        }

        case 'S': {
            value = 'Subdivision';
            break;
        }

        default: {
            value = documentType
            break;
        }
    }

    return value;
}

function dateToYear(date: string) {

    if (!date || !new Date(date).getFullYear()) {
        return
    }

    return new Date(date).getFullYear();
}

function arrayObjectKeyToString(value, options) {

    const { path, separator } = options;
    if (!Array.isArray(value)) {
        throw new Error('the value is not array');
    }

    return value
        .filter(entry => get(entry, path))
        .map(entry => get(entry, path))
        .join(separator);
}

function arrayObjectKeyToArrayString(value, options) {

    const { path, separator } = options;
    if (!Array.isArray(value)) {
        throw new Error('the value is not array');
    }

    return value
        .filter(entry => get(entry, path))
        .map(entry => get(entry, path));
}

function stringArrayToObjectArray(stringArrayValue: string[], options: { targetPath: string }): (Record<string, any> | undefined)[] | undefined {
    if (!stringArrayValue || !Array.isArray(stringArrayValue)) {
        return undefined
    }
    return stringArrayValue.map(str => {
        const original = {}
        objectPath.set(original, options.targetPath, str)
        return original
    })
}

function companyAddressType(addressType: string) {
    let value;

    switch (addressType) {
        case 'HQ': {
            value = 'HEADQUARTERS';
            break;
        }

        case 'AGENT ADDRESS': {
            value = 'OFFICER';
            break;
        }

        default: {
            value = addressType.toLowerCase().includes('officer') ?
                'OFFICER' :
                'OTHER'
            break;
        }
    }

    return value;
}

function fieldConditionMapping(value, options) {

    const {
        pathToCheck,
        condition,
        conditionValue,
        valueToMap,
        pathToMap,
        keepOriginalValue
    }: {
        pathToCheck: string,
        condition: 'exists' | 'equals' | 'contains' | 'notEqual' | 'notContain',
        conditionValue: string,
        valueToMap: string | number,
        pathToMap: string,
        keepOriginalValue: boolean
    } = options;

    const valueToCheck: string = get(value, pathToCheck);


    let pass;

    switch (condition) {
        case 'exists':
            const flag = valueToCheck !== undefined && valueToCheck !== null;
            pass = conditionValue ? flag : !flag;
            break;
        case 'equals':
            pass = ((valueToCheck !== undefined && valueToCheck !== null) && conditionValue === valueToCheck);
            break;
        case 'contains':
            pass = ((valueToCheck !== undefined && valueToCheck !== null) && (valueToCheck || '').toLowerCase().includes(conditionValue.toLowerCase()));;
            break;
        case 'notEqual':
            pass = (valueToCheck && !(conditionValue === valueToCheck));
            break;
        case 'notContain':
            pass = ((valueToCheck !== undefined && valueToCheck !== null) && !(valueToCheck || '').toLowerCase().includes(conditionValue.toLowerCase()));;
            break;
    }

    return pass ?
        pathToMap ?
            get(value, pathToMap) :
            valueToMap :
        keepOriginalValue ? value : undefined
}

function invertBooleanValue(value: boolean) {
    if (typeof value !== 'boolean') {
        return undefined
    }
    return !value
}


function joinObjectKeysToString(value, options) {
    const {
        keysToJoin,
        seperator
    }: {
        keysToJoin: string[],
        seperator: string
    } = options;

    let valueToMap = '';

    for (const key of keysToJoin) {
        const valueToCheck: string = get(value, key);
        if (valueToCheck) {
            if (seperator) {
                valueToMap = valueToMap ? valueToMap + `${seperator} ${valueToCheck}` : valueToCheck;
                continue;
            }
            valueToMap = valueToMap ? valueToMap + ` ${valueToCheck}` : valueToCheck;
        }
        valueToMap = valueToMap.trim();
    }
    return valueToMap;
}

function companyNameTransformer(str: string) {
    if (!str) {
        return str;
    }

    const businessStructureAbbreviations = [
        "Col", "Corp", "Corp.", "Inc", "LC", "LLC", "LLLP", "LLP", "LP", "Ltd", "PC", "PLLC", "GP",
        "Co.", "col", "corp", "corp.", "inc", "lc", "llc", "lllp", "llp", "lp", "ltd", "pc", "pllc", "gp",
        "co"];


    const words = str.trim().split(/\s+/);

    const transformedWords = words.map(word => {
        const cleanedWord = word.replace(/^[^a-zA-Z0-9]+|[^a-zA-Z0-9]+$/g, '');

        if (businessStructureAbbreviations.includes(cleanedWord)) {
            return word;
        }

        return titleCaseTransformer(word);
    });

    return transformedWords.join(' ');
}

async function convertFastCaseDate(fastCaseDate: string) {

    if (!fastCaseDate) {
        return;
    }

    const jsDate = new Date(fastCaseDate);

    if (jsDate instanceof Date && !isNaN(jsDate.getTime())) {
        return jsDate;
    }

    const match = fastCaseDate.match(/\/Date\((-?\d+)([+-]\d{4})\)\//);

    if (!match) {
        throw new Error('Invalid date format');
    }
    const milliseconds = parseInt(match[1], 10);
    const offset = match[2];

    const date = new Date(milliseconds);

    const offsetHours = parseInt(offset.slice(0, 3), 10);
    const offsetMinutes = parseInt(offset.slice(3), 10);
    const offsetMilliseconds = (offsetHours * 60 + offsetMinutes) * 60 * 1000;

    const normalizedDate = new Date(date.getTime() - offsetMilliseconds);

    return normalizedDate;
}

function parseClearDate(clearDate: string): Date {

    const cleanedInput = clearDate.replace(/^(ca\.?|around|approximately)\s*/i, '').trim();
    if (String(cleanedInput).length === 4) {
        return new Date(Number(clearDate), 0, 1)
    }

    return chrono.parseDate(clearDate) as Date;
}

function convertClearDate(clearDate: string) {

    if (!clearDate) {
        return;
    }

    if (clearDate.includes(';')) {

        const dates = clearDate.split(';');


        if (dates.length === 1) {
            return parseClearDate(dates[0]);
        }

        const [earlyDate] = dates
            .map(date => date.trim())
            .map(date => parseClearDate(date))
            .sort((a, b) => a!.getTime() - b!.getTime());

        return earlyDate;
    }

    return parseClearDate(clearDate);
}

function parseSteeleDate(steeleDate: string): Date {

    if (/^\d{2}\.\d{2}\.\d{4}$/.test(steeleDate)) {
        const [day, month, year] = steeleDate.split('.');
        return new Date(`${year}-${month}-${day}`);
    }

    if (/^\d{4}-\d{2}$/.test(steeleDate)) {
        return new Date(`${steeleDate}-01`); // Add day '01'
    }

    return new Date(`${steeleDate}-01-01`); // Add month '01' and day '01'
}

function convertSteeleDate(steeleDate: string) {

    if (!steeleDate) {
        return;
    }

    if (steeleDate.includes(';')) {

        const dates = steeleDate.split(';');


        if (dates.length === 1) {
            return parseSteeleDate(steeleDate[0]);
        }

        const [earlyDate] = dates
            .map(date => date.trim())
            .map(date => parseSteeleDate(date))
            .sort((a, b) => a!.getTime() - b!.getTime());

        return earlyDate;
    }

    return parseSteeleDate(steeleDate);
}

export const strategies = {
    predefinedTransformations: {
        toUpperCase: (str: string | string[]) => toUpperCase(str),
        toLowerCase: (str: string | string[]) => toLowerCase(str),
        titleCase: (str: string | string[]) => titleCase(str),
        companyNameFormat: (str: string | string[]) => companyNameFormat(str),
        toDate: (str: string | string[]) => toDate(str),
        arrayToString: (arr: string[], separator: string) => arr.join(separator),
        stringToArray: (str: string | string[], separator: string) => stringToArray(str, separator),
        fromTracerDate,
        tracerPropertyDocumentType,
        tracerPropertyTransferType,
        companyAddressType,
        dateToYear,
        arrayObjectKeyToString,
        arrayObjectKeyToArrayString,
        fieldConditionMapping,
        joinObjectKeysToString,
        invertBooleanValue,
        stringArrayToObjectArray,
        convertStringToDate: (str: string) => chrono.parseDate(str),
        convertFastCaseDate,
        convertClearDate,
        convertSteeleDate
    }
}