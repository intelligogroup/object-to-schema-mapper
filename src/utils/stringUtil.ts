function unidotify(str: string) {
    const unicodeDot = '\u2024';
    return replaceAll('.', unicodeDot)(str);
}

function unUnidotify(str: string) {
    const unicodeDot = '\u2024';
    return replaceAll(unicodeDot, '.')(str);
}

function replaceAll(original: string, substitute: string) {
    return (str: string) => str.split(original).join(substitute);
}

export {
    unidotify,
    unUnidotify,
    replaceAll
}