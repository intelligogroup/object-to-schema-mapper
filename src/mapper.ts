import objectPath from 'object-path';
import R from 'ramda';
import { ITarget } from './utils/types';
import { SomeObj, Transform, TreeLeaf, } from './utils/types';
import { unidotify, unUnidotify } from './utils/stringUtil';
import { unique, identity } from './utils/generalUtil';
import { postProcessCreatedObject } from './postProcessing';
import {
    generateTransform,
    treeLeafsToTransforms,
    groupTransformsByTarget,
    sortTransformsByPriority,
    pruneEmpty,
} from './utils/transform';
import { strategies } from './utils/predefinedTransformations';

function mapObject(originalObj: SomeObj, transformations: Transform[]) {
    if (!transformations || transformations.length == 0) {
        return {};
    }
    const initialObject = createTargetObject(transformations, originalObj);
    const postProcessedObject = postProcessCreatedObject(initialObject, transformations);
    return pruneEmpty(postProcessedObject);
}

function createTargetObject(
    transformations: Transform[],
    originalObj: SomeObj,
    targetObject: SomeObj = {}
) {
    const extractionTree = buildTree(transformations);
    const [treeEntries, treeLeafs] = R.partition(isTargetValue(), Object.entries(extractionTree));

    treatLeafsMutation(originalObj, treeLeafs, targetObject);
    treatTreeMutation(originalObj, treeEntries, transformations, targetObject);

    return targetObject;
}

function treatTreeMutation(originalObj: SomeObj, treeEntries: any[], transformations: Transform[], targetObject: SomeObj) {
    treeEntries.forEach(([source, _]) => {

        const arrayValuesToSet = objectPath.get(originalObj, source);

        if (!arrayValuesToSet) {
            return
        }

        const nextTargets = getSubTransformations(transformations, source);
        nextTargets.forEach(({ superTarget, transforms }) => {

            arrayValuesToSet.forEach(originalSubObj => {

                objectPath.push(targetObject, superTarget, createTargetObject(transforms, originalSubObj));
            });
        });
    });
}

function chooseHighestPriorityTransforms(originalObj: SomeObj) {
    return function withObject(transforms: Transform[]): Transform[] {
        if (transforms.every(transform => transform.target.path.includes('[]'))) {
            return transforms;
        }

        const transformGenerator = generateTransform(transforms);
        let transform = transformGenerator.next();
        let valueToSet = objectPath.get(originalObj, unUnidotify(transform.value.source)) || transform.value.target.defaultValue || null;

        while (
            !transform.done
            && !(objectPath.has(originalObj, unUnidotify(transform.value.source))
                && (!transform.value.target.predefinedTransformations?.length
                    || getValueAfterTransformations(transform.value.target, valueToSet)))
            && !transform.value.target.defaultValue
        ) {
            
            transform = transformGenerator.next();
            
            if (transform.done) {
                continue;
            }

            valueToSet = objectPath.get(originalObj, unUnidotify(transform.value.source)) || transform.value.target.defaultValue || null;
        }
        return transform.done ? [] : [transform.value];
    }
}

function transformsToTreeLeafs(transforms: Transform[]): TreeLeaf[] {
    const grouped = R.groupBy((transform: Transform) => transform.source, transforms) as Partial<Record<string, Transform[]>>;

    return Object
        .entries(grouped)
        .map(([path, transforms]) => [path, (transforms ?? []).map(transform => transform.target)]);
}

function eliminateLowPriority(originalObj: SomeObj, treeLeafs: TreeLeaf[]): TreeLeaf[] {
    const transforms = treeLeafsToTransforms(treeLeafs);
    const tranformsByTarget = groupTransformsByTarget(transforms);
    const highestPriorityTransforms = Object
        .entries(tranformsByTarget)
        .map(([path, transforms]) => [
            path,
            chooseHighestPriorityTransforms(originalObj)(sortTransformsByPriority(transforms))
        ])
        .filter(([_, transforms]) => (transforms as Transform[]).length)
        .flatMap(([_, transforms]) => (transforms as Transform[]).map(identity));
    const priorityTreeLeafs = transformsToTreeLeafs(highestPriorityTransforms as Transform[]);
    return priorityTreeLeafs;
}

function getValueAfterTransformations(target: ITarget, valueToSet: SomeObj) {
    return target?.predefinedTransformations
        ?.reduceRight(
            (finalValue, predefinedTransformation) => {
                const transformationName = predefinedTransformation?.transformation;
                const transformationArgs = predefinedTransformation?.transformationArgs;
                const options = predefinedTransformation?.options;

                return (strategies.predefinedTransformations[transformationName] ?? identity)(finalValue, transformationArgs || options);
            },
            valueToSet
        );
}

function treatLeafsMutation(originalObj: SomeObj, treeLeafs: TreeLeaf[], targetObject: SomeObj) {
    const priorityTreeLeafs = eliminateLowPriority(originalObj, treeLeafs);
    priorityTreeLeafs.forEach(([source, mappingsArray]) => {
        mappingsArray.forEach(target => {
            const value = objectPath.get(originalObj, unUnidotify(source));
            let valueToSet = value === 0 ? value : (value || target.defaultValue || null);
            if (target.predefinedTransformations) {
                valueToSet = getValueAfterTransformations(target, valueToSet)
            }
            if (target.path.includes('[]')) {
                const splitTargetByArray = target.path.split('[]');
                const valueToPush = R.isEmpty(splitTargetByArray[1])
                    ? valueToSet
                    : { [splitTargetByArray[1].slice(1)]: valueToSet };
                objectPath.push(targetObject, splitTargetByArray[0], valueToPush);
            } else {
                objectPath.set(targetObject, target.path, valueToSet);
            }
        });
    });
}

function getSubTransformations(transformations: Transform[], immediateSource: string) {
    const immediateTargets = transformations
        .filter(transform => transform.source.split('[].')[0] == immediateSource)
        .map(transform => {
            const sourceSplit = transform.source.split('[].')
            const targetSplit = transform.target.path.split('[].');
            const idxOfNesting = sourceSplit.indexOf(immediateSource);
            return {
                superTarget: targetSplit[idxOfNesting],
                innerTransformation: {
                    source: sourceSplit.slice(idxOfNesting + 1).join('[].'),
                    target: {
                        ...transform.target,
                        path: targetSplit.slice(idxOfNesting + 1).join('[].'),
                    }
                }
            };
        })
        .filter(unique);

    const groupedTransforms = R.groupBy(target => target.superTarget, immediateTargets);
    return Object
        .entries(groupedTransforms)
        .map(([superTarget, value]) => ({
            superTarget,
            transforms: value!.map(el => el.innerTransformation)
        }));
}

function isTargetValue() {
    return ([_, value]) => !Array.isArray(value);
}

function buildTree(transformations: Transform[]): SomeObj {
    const { group: groupedTransforms, leafs: leafTransforms } = groupExtractionsByPath(transformations);
    return Object
        .entries(groupedTransforms)
        .reduce(
            (acc, [key, value]) => ({
                ...acc,
                ...{
                    [key]: buildTree(
                        value.map(transform => ({
                            source: transform.source.split(`${key}[].`)[1],
                            target: transform.target
                        }))
                    )
                }
            }),
            leafTransforms.reduce(
                (acc, leaf) => {
                    objectPath.push(acc, unidotify(leaf.source), leaf.target);
                    return acc;
                },
                {}
            )
        );
}

function groupExtractionsByPath(transformations: Transform[]): { group: SomeObj, leafs: Transform[] } {
    const [furtherProcessing, noProcessing] = R.partition(transform => transform.source.includes('[]'), transformations);
    return {
        group: groupBySourcePrefix()(furtherProcessing),
        leafs: noProcessing
    }
}

function groupBySourcePrefix() {
    return R.groupBy((transform: Transform) => transform.source.split('[]')[0]);
}

export { mapObject }
