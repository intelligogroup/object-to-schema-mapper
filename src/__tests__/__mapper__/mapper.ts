import data from './data.json';
import { transformations } from './transformations.json';

import { mapObject } from '../../mapper';
import { Transform } from '../../utils/types';

describe('transform data object into target object according rules in transformations', () => {

    test('creates an object that is the result of applying all transformations on the data object', () => {
        expect(mapObject(data, transformations as Transform[])).toMatchSnapshot()
    })

})
