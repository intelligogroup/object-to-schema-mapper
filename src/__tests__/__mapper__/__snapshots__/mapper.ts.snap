// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform data object into target object according rules in transformations creates an object that is the result of applying all transformations on the data object 1`] = `
Object {
  "addresses": Array [
    Object {
      "houseNumber": "h1",
      "streetAddress": "s1",
    },
    Object {
      "houseNumber": "h2",
      "streetAddress": "s2",
    },
    Object {
      "houseNumber": "h1B",
      "streetAddress": "wall street",
    },
    Object {
      "houseNumber": "h2B",
      "streetAddress": "s2B",
    },
  ],
  "d": Object {
    "Dates": Array [
      "1/1/66",
      "10/14/2026",
    ],
  },
  "dates": Array [
    Object {
      "dob": "1/1/66",
    },
    Object {
      "dod": "10/14/2026",
    },
  ],
  "dob": "1/1/66",
  "jobs": Object {
    "currentPosition": "e",
    "lastPositions": "A, B, C, D",
  },
  "names": Object {
    "first": "john",
    "last": "doe",
    "nicknames": Array [
      "John",
      "Johnny",
      "John-boy",
      "BigJ",
      "Bbj",
      "Mj",
    ],
  },
  "other": Array [
    Object {
      "transformedId": 1,
    },
    Object {
      "transformedId": 2,
    },
    Object {
      "transformedId": 3,
    },
  ],
  "some": Object {
    "time": Array [
      Object {
        "test": Array [
          Object {
            "no": "f1",
            "p": 1,
          },
          Object {
            "no": "f2",
            "p": 2,
          },
        ],
      },
      Object {
        "test": Array [
          Object {
            "no": "f3",
            "p": 3,
          },
          Object {
            "no": "f4",
            "p": 4,
          },
        ],
      },
      Object {
        "test": Array [
          Object {
            "no": "f5",
            "p": 5,
          },
          Object {
            "no": "f6",
            "p": 6,
          },
        ],
      },
    ],
  },
}
`;
